code/controllers/subsystem/tts.dm

```diff
@@ -0,0 +1,207 @@
+	request.begin_async()
+	in_process_tts_messages += list(entry)
+
+/datum/controller/subsystem/tts/fire(resumed)
+	if(!tts_enabled)
+		flags |= SS_NO_FIRE
+		return
+
+	if(!resumed)
+		while(length(in_process_tts_messages) < max_concurrent_requests && queued_tts_messages.len > 0)
+			var/list/entry = popleft(queued_tts_messages)
+			handle_request(entry)
+		current_processing_tts_messages = in_process_tts_messages.Copy()
+
+	// For speed
+	var/list/processing_messages = current_processing_tts_messages
+	while(processing_messages.len)
+		var/current_message = processing_messages[processing_messages.len]
+		processing_messages.len--
+		if(current_message[TIMEOUT_INDEX] < world.time)
+			in_process_tts_messages -= list(current_message)
+			cached_voices -= current_message[IDENTIFIER_INDEX]
+			continue
+
+		var/datum/http_request/request = current_message[REQUEST_INDEX]
+		if(!request.is_complete())
+			continue
+
+		var/datum/http_response/response = request.into_response()
+		in_process_tts_messages -= list(current_message)
+		if(response.errored)
+			cached_voices -= current_message[IDENTIFIER_INDEX]
+			continue
+		var/identifier = current_message[IDENTIFIER_INDEX]
+		var/sound/new_sound = new("tmp/tts/[identifier].ogg")
+		play_tts(current_message[TARGET_INDEX], new_sound, current_message[LANGUAGE_INDEX], current_message[LOCAL_INDEX])
+		for(var/extra_target in current_message[EXTRA_TARGETS_INDEX])
+			play_tts(extra_target["target"], new_sound, current_message[LANGUAGE_INDEX], extra_target["local"])
+		cached_voices -= identifier
+		if(MC_TICK_CHECK)
+			return
+
+/datum/controller/subsystem/tts/proc/queue_tts_message(target, message, datum/language/language, speaker, filter, local = FALSE)
+	if(!tts_enabled)
+		return
+
+	var/static/regex/contains_alphanumeric = regex("\[a-zA-Z0-9]", "g")
+	// If there is no alphanumeric char, the output will usually be static, so
+	// don't bother sending
+	if(contains_alphanumeric.Find(message) == 0)
+		return
+
+	var/shell_scrubbed_input = tts_alphanumeric_filter(message)
+	shell_scrubbed_input = copytext(shell_scrubbed_input, 1, 300)
+	var/identifier = sha1(speaker + shell_scrubbed_input + filter)
+	var/cached_voice = cached_voices[identifier]
+	if(islist(cached_voice))
+		cached_voice[EXTRA_TARGETS_INDEX] += list(list(target = target, local = local))
+		return
+	else if(fexists("tmp/tts/[identifier].ogg"))
+		var/sound/new_sound = new("tmp/tts/[identifier].ogg")
+		play_tts(target, new_sound, language, local)
+		return
+	if(!(speaker in available_speakers))
+		return
+	speaker = tts_alphanumeric_filter(speaker)
+
+	var/list/headers = list()
+	headers["Content-Type"] = "application/json"
+	var/datum/http_request/request = new()
+	var/file_name = "tmp/tts/[identifier].ogg"
+	request.prepare(RUSTG_HTTP_METHOD_GET, "[CONFIG_GET(string/tts_http_url)]/tts?voice=[speaker]&identifier=[identifier]&filter=[url_encode(filter)]", json_encode(list("text" = shell_scrubbed_input)), headers, file_name)
```

Comment:
Since we are going to be using a tts gpu instance, these requests will have to go out over the internet, and as such i'll need the ability to add an auth token to this.

Reply:
alright